test/integration/framework/util.go

```diff
@@ -43,3 +64,227 @@ func CreateTestingNamespace(baseName string, apiserver *httptest.Server, t *test
+	return isNodeConditionSetAsExpected(node, conditionType, wantTrue, true)
+}
+
+// isConditionUnset returns true if conditions of the given node do not have a match to the given conditionType, otherwise false.
+func isConditionUnset(node *v1.Node, conditionType v1.NodeConditionType) bool {
+	for _, cond := range node.Status.Conditions {
+		if cond.Type == conditionType {
+			return false
+		}
+	}
+	return true
+}
+
+func isNodeConditionSetAsExpected(node *v1.Node, conditionType v1.NodeConditionType, wantTrue, silent bool) bool {
+	// Check the node readiness condition (logging all).
+	for _, cond := range node.Status.Conditions {
+		// Ensure that the condition type and the status matches as desired.
+		if cond.Type == conditionType {
+			// For NodeReady condition we need to check Taints as well
+			if cond.Type == v1.NodeReady {
+				hasNodeControllerTaints := false
+				// For NodeReady we need to check if Taints are gone as well
+				taints := node.Spec.Taints
+				for _, taint := range taints {
+					if taint.MatchTaint(nodectlr.UnreachableTaintTemplate) || taint.MatchTaint(nodectlr.NotReadyTaintTemplate) {
+						hasNodeControllerTaints = true
+						break
+					}
+				}
+				if wantTrue {
+					if (cond.Status == v1.ConditionTrue) && !hasNodeControllerTaints {
+						return true
+					}
+					msg := ""
+					if !hasNodeControllerTaints {
+						msg = fmt.Sprintf("Condition %s of node %s is %v instead of %t. Reason: %v, message: %v",
+							conditionType, node.Name, cond.Status == v1.ConditionTrue, wantTrue, cond.Reason, cond.Message)
+					} else {
+						msg = fmt.Sprintf("Condition %s of node %s is %v, but Node is tainted by NodeController with %v. Failure",
+							conditionType, node.Name, cond.Status == v1.ConditionTrue, taints)
+					}
+					if !silent {
+						klog.Infof(msg)
+					}
+					return false
+				}
+				// TODO: check if the Node is tainted once we enable NC notReady/unreachable taints by default
+				if cond.Status != v1.ConditionTrue {
+					return true
+				}
+				if !silent {
+					klog.Infof("Condition %s of node %s is %v instead of %t. Reason: %v, message: %v",
+						conditionType, node.Name, cond.Status == v1.ConditionTrue, wantTrue, cond.Reason, cond.Message)
+				}
+				return false
+			}
+			if (wantTrue && (cond.Status == v1.ConditionTrue)) || (!wantTrue && (cond.Status != v1.ConditionTrue)) {
+				return true
+			}
+			if !silent {
+				klog.Infof("Condition %s of node %s is %v instead of %t. Reason: %v, message: %v",
+					conditionType, node.Name, cond.Status == v1.ConditionTrue, wantTrue, cond.Reason, cond.Message)
+			}
+			return false
+		}
+
+	}
+	if !silent {
+		klog.Infof("Couldn't find condition %v on node %v", conditionType, node.Name)
+	}
+	return false
+}
+
+// isNodeUntainted tests whether a fake pod can be scheduled on "node", given its current taints.
+// TODO: need to discuss wether to return bool and error type
+func isNodeUntainted(node *v1.Node) bool {
```

Comment:
Why do we need to implement `isNodeUntaintedWithNonblocking()`?
The function is only called from this function, and the above seems unnecessary wrapper.
We can change like
```
- func isNodeUntaintedWithNonblocking(node *v1.Node, nonblockingTaints string) bool {
+ func isNodeUntainted(node *v1.Node) bool {
+     nonblockingTaints  := ""
```

Reply:
Make sense. Honestly, the changes in this PR make me uncomfortable. You know, lots of duplicated code across the codebase, especially with lots of TODO items. Maybe we should rethink about this one https://github.com/kubernetes/kubernetes/blob/42972f2a37c56b5f850086a7330b34ae6592f8ac/test/integration/framework/util.go#L17.

Reply:
@gavinfish 
Many TODOs cannot be traced to their causes or historical background due to code changes, movements, and refactorings. Welcome to solve some of them, haha
see  https://github.com/kubernetes/kubernetes/issues/86763  https://github.com/kubernetes/kubernetes/issues/86756 
welcome

Reply:
> Make sense. Honestly, the changes in this PR make me uncomfortable. You know, lots of duplicated code across the codebase, especially with lots of TODO items. Maybe we should rethink about this one 

I agree with this. Little duplication here and there is okay but this is a big one. Wondering if there's a cleaner way to do this.

Reply:
@andrewsykim I have created #89292 to track this issue.